### ISSUE TITLE
Analysis and enforcement of business logic: Climsoft `stationelement`

### ISSUE THREAD

**Issue by isedwards** [opencdms-dev/pyopencdms-old#45]
The Climsoft `stationelement` table is intended to maintain a correct record of which elements are recorded at which stations - including the beginning and end dates for available data.

For these types of metadata tables pyopencdms should have access to the following processes:
1. A process for checking that the data currently contained in the `stationelement` table is correct and complete - and updating the table where necessary
2. A process for updating the table when required due to data and metadata changes in the database (further information to be added in comments below)

Below is SQL DDL code describing the `stationelement` table:

```
CREATE TABLE `stationelement` (
	`recordedFrom` VARCHAR(255) NULL DEFAULT NULL COLLATE 'latin1_swedish_ci',
	`describedBy` BIGINT(20) NULL DEFAULT NULL,
	`recordedWith` VARCHAR(255) NULL DEFAULT NULL COLLATE 'latin1_swedish_ci',
	`instrumentcode` VARCHAR(6) NULL DEFAULT NULL COLLATE 'latin1_swedish_ci',
	`scheduledFor` VARCHAR(255) NULL DEFAULT NULL COLLATE 'latin1_swedish_ci',
	`height` FLOAT(6,2) NULL DEFAULT NULL,
	`beginDate` VARCHAR(50) NULL DEFAULT NULL COLLATE 'latin1_swedish_ci',
	`endDate` VARCHAR(50) NULL DEFAULT NULL COLLATE 'latin1_swedish_ci',
	UNIQUE INDEX `stationElementIdentification` (`recordedFrom`, `describedBy`, `recordedWith`, `beginDate`) USING BTREE,
	INDEX `obsElementobservationInitial` (`describedBy`) USING BTREE,
	INDEX `stationobservationInitial` (`recordedFrom`) USING BTREE,
	INDEX `FK_mysql_climsoft_db_v4_elementRecordedWith` (`recordedWith`) USING BTREE,
	INDEX `FK_mysql_climsoft_db_v4_elementScheduledFor` (`scheduledFor`) USING BTREE,
	CONSTRAINT `FK_mysql_climsoft_db_v4_elementDescribedBy` FOREIGN KEY (`describedBy`) REFERENCES `mariadb_climsoft_test_db_v4`.`obselement` (`elementId`) ON UPDATE RESTRICT ON DELETE RESTRICT,
	CONSTRAINT `FK_mysql_climsoft_db_v4_elementRecordedWith` FOREIGN KEY (`recordedWith`) REFERENCES `mariadb_climsoft_test_db_v4`.`instrument` (`instrumentId`) ON UPDATE RESTRICT ON DELETE RESTRICT,
	CONSTRAINT `FK_mysql_climsoft_db_v4_elementScheduledFor` FOREIGN KEY (`scheduledFor`) REFERENCES `mariadb_climsoft_test_db_v4`.`obsscheduleclass` (`scheduleClass`) ON UPDATE RESTRICT ON DELETE RESTRICT,
	CONSTRAINT `FK_mysql_climsoft_db_v4_staionRecordedFrom` FOREIGN KEY (`recordedFrom`) REFERENCES `mariadb_climsoft_test_db_v4`.`station` (`stationId`) ON UPDATE RESTRICT ON DELETE RESTRICT
)
COLLATE='latin1_swedish_ci'
ENGINE=InnoDB
;

```

**Comment by faysal-ishtiaq** [comment]
@isedwards I am listing the key-points of our discussion on this issue here:

1. Use recordedFrom, describedBy and obsDatetime from observationfinal to populate recordedFrom, describedBy and beginDate respectively.
2.  if data of that type hasn't been recorded for a particular period ( e.g. 2 months) it would be good to have an option to automatically close it (using the date of the most recent observation)

**Comment by isedwards** [comment]
Resolved by #46
